File: src/lexaloud/daemon.py
"""FastAPI daemon exposing the Lexaloud HTTP API on loopback.

Routes:
    POST /speak  {text, mode?}
    POST /pause
    POST /resume
    POST /stop
    POST /skip
    POST /back
    GET  /state
    GET  /healthz

Design notes:
- Uses a dedicated ThreadPoolExecutor for provider work so that interpreter
  shutdown (triggered by systemd SIGTERM) can bail out without waiting on
  the default executor's non-daemon worker threads.
- The provider's warmup runs as a background task during lifespan; while
  warmup runs, the player's state is reported as "warming" via
  `player.set_warming(True)`.
- The /speak middleware enforces a Content-Length cap using JSONResponse
  directly — HTTPException raised from middleware does NOT pass through
  FastAPI's exception handlers, so returning the response here is essential.
"""

from __future__ import annotations

import asyncio
import concurrent.futures
import logging
from contextlib import asynccontextmanager
from dataclasses import asdict, dataclass
from typing import Any, Literal

from fastapi import FastAPI, HTTPException, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from . import __version__
from .audio import AudioSink, SoundDeviceSink
from .config import Config, load_config, runtime_dir, socket_path
from .models import assert_onnxruntime_environment, ensure_artifacts
from .player import Player, PlayerState
from .preprocessor import PreprocessorConfig, preprocess_with_llm
from .providers.base import SpeechProvider
from .providers.kokoro import KokoroProvider

log = logging.getLogger(__name__)


# Hard cap on the length of a single sentence after preprocessing. Dense
# academic prose can legitimately hit 500-1500 characters; 4096 is a
# generous ceiling that still rules out pathological input (e.g., an
# uninterrupted 200 KB paragraph with no sentence boundaries, which
# would either OOM the GPU or silently produce garbage audio).
MAX_SENTENCE_CHARS = 4096


# ---------- request/response models ----------


class SpeakRequest(BaseModel):
    text: str = Field(..., min_length=1)
    mode: Literal["replace", "append"] = "replace"


class StateResponse(BaseModel):
    state: str
    current_sentence: str | None
    pending_count: int
    ready_count: int
    provider_name: str
    session_providers: list[str]
    last_error: str | None = None


# ---------- daemon wiring ----------


@dataclass
class DaemonComponents:
    cfg: Config
    provider: SpeechProvider
    sink: AudioSink
    player: Player
    preproc_config: PreprocessorConfig
    normalizer: Any = None  # LlmNormalizer | None


def build_components(cfg: Config | None = None) -> DaemonComponents:
    cfg = cfg or load_config()

    # Runtime guard: refuse to start if the ONNX Runtime environment is broken.
    ort_distribution = assert_onnxruntime_environment()

    artifacts = ensure_artifacts(download_if_missing=False)
    provider = KokoroProvider(
        model_path=artifacts["kokoro-v1.0.onnx"],
        voices_path=artifacts["voices-v1.0.bin"],
        voice=cfg.provider.voice,
        lang=cfg.provider.lang,
        speed=cfg.provider.speed,
        # The CPU AppImage contains the CPU distribution by design.  Select
        # the CUDA path only when the installed ORT distribution can provide
        # it; otherwise skip CUDA probing entirely instead of emitting a
        # misleading fallback warning on every startup.
        prefer_cuda=ort_distribution == "onnxruntime-gpu",
    )
    sink: AudioSink = SoundDeviceSink()
    player = Player(provider=provider, sink=sink, ready_queue_depth=cfg.daemon.ready_queue_depth)
    preproc_config = PreprocessorConfig(
        dedupe_mathjax_selection=cfg.preprocessor.dedupe_mathjax_selection,
        strip_markdown=cfg.preprocessor.strip_markdown,
        sre_latex_enabled=cfg.sre_latex.enabled,
        sre_latex_timeout_s=cfg.sre_latex.timeout_s,
        sre_latex_domain=cfg.sre_latex.domain,
        sre_latex_style=cfg.sre_latex.style,
        strip_numeric_bracket_citations=cfg.preprocessor.strip_numeric_bracket_citations,
        strip_parenthetical_citations=cfg.preprocessor.strip_parenthetical_citations,
        expand_latin_abbreviations=cfg.preprocessor.expand_latin_abbreviations,
        expand_academic_abbreviations=cfg.preprocessor.expand_academic_abbreviations,
        normalize_numbers=cfg.preprocessor.normalize_numbers,
        normalize_urls=cfg.preprocessor.normalize_urls,
        normalize_math_symbols=cfg.preprocessor.normalize_math_symbols,
        pdf_cleanup=cfg.preprocessor.pdf_cleanup,
    )
    normalizer = None
    if cfg.normalizer.enabled:
        try:
            from .preprocessor.llm_normalize import LlmNormalizer

            normalizer = LlmNormalizer(cfg.normalizer)
        except Exception as e:  # noqa: BLE001
            log.warning("LLM normalizer initialization failed: %s", e)

    return DaemonComponents(
        cfg=cfg,
        provider=provider,
        sink=sink,
        player=player,
        preproc_config=preproc_config,
        normalizer=normalizer,
    )


def create_app(components: DaemonComponents | None = None) -> FastAPI:
    comps = components or build_components()

    @asynccontextmanager
    async def lifespan(app: FastAPI):
        # Use a dedicated ThreadPoolExecutor so shutdown can bail out without
        # waiting on the default executor's non-daemon worker threads. This
        # makes systemd SIGTERM responsive even if a Kokoro.create() is in
        # flight in the executor at the moment of shutdown.
        exec_pool = concurrent.futures.ThreadPoolExecutor(
            max_workers=2, thread_name_prefix="lexaloud-worker"
        )
        loop = asyncio.get_running_loop()
        loop.set_default_executor(exec_pool)

        comps.player.set_warming(True)

        async def _warmup_bg() -> None:
            try:
                await comps.provider.warmup()
            except Exception as e:  # noqa: BLE001
                log.error("background warmup failed: %s", e)
            # Sink warmup: pre-open the audio stream so PipeWire's
            # 24000→44100 Hz resampler is initialized before the user's
            # first hotkey press. Best-effort — the daemon works without
            # it (begin_stream cold-opens lazily), but the first sentence
            # would clip without the warm stream.
            try:
                await comps.sink.warmup(24000, 1)
            except Exception as e:  # noqa: BLE001
                log.warning("sink warmup failed (audio device may be unavailable): %s", e)
            # LLM normalizer warmup (optional) [M9]
            if comps.normalizer is not None:
                try:
                    await comps.normalizer.warmup()
                except Exception as e:  # noqa: BLE001
                    log.warning("LLM normalizer warmup failed: %s", e)
            comps.player.set_warming(False)

        warmup_task = asyncio.create_task(_warmup_bg(), name="lexaloud-warmup")

        # MPRIS2 integration (optional — dbus-fast may not be installed)
        mpris = None
        try:
            from .mpris import MprisAdapter

            mpris = MprisAdapter(comps.player, comps.cfg)
            await mpris.connect()
        except Exception as e:  # noqa: BLE001
            log.warning("MPRIS2 unavailable: %s", e)
            mpris = None

        # XDG GlobalShortcuts portal (KDE/Sway/Hyprland; NOT GNOME)
        shortcuts = None
        try:
            from .shortcuts import ShortcutsAdapter

            shortcuts = ShortcutsAdapter(
                comps.player, comps.cfg, comps.preproc_config, comps.normalizer
            )
            registered = await shortcuts.try_register()
            if not registered:
                shortcuts = None
        except Exception as e:  # noqa: BLE001
            log.info("GlobalShortcuts portal unavailable: %s", e)
            shortcuts = None

        try:
            yield
        finally:
            if shortcuts is not None:
                shortcuts.disconnect()
            if mpris is not None:
                try:
                    mpris.disconnect()
                except Exception:  # noqa: BLE001
                    pass
            warmup_task.cancel()
            try:
                await warmup_task
            except (asyncio.CancelledError, Exception):  # noqa: BLE001
                pass
            await comps.player.shutdown()
            if comps.normalizer is not None:
                comps.normalizer.shutdown()
            exec_pool.shutdown(wait=False, cancel_futures=True)

    app = FastAPI(title="Lexaloud", version=__version__, lifespan=lifespan)
    app.state.components = comps

    max_bytes = comps.cfg.capture.max_bytes
    # Allow a small overhead for JSON envelope (keys + quotes + escapes).
    HARD_HEADER_CAP = max_bytes + 4096

    @app.middleware("http")
    async def _payload_guard(request: Request, call_next):
        # HTTPException raised from middleware does NOT pass through
        # FastAPI's exception handler machinery — it surfaces as a 500.
        # We must return a JSONResponse directly.
        if request.url.path == "/speak":
            content_length = request.headers.get("content-length")
            if content_length is not None:
                try:
                    if int(content_length) > HARD_HEADER_CAP:
                        return JSONResponse(
                            status_code=413,
                            content={"detail": "payload too large"},
                        )
                except ValueError:
                    pass
        return await call_next(request)

    @app.get("/healthz")
    async def healthz() -> dict:
        return {"status": "ok"}

    @app.get("/state", response_model=StateResponse)
    async def get_state() -> StateResponse:
        s: PlayerState = comps.player.state
        return StateResponse(**asdict(s))

    @app.post("/speak", response_model=StateResponse)
    async def speak(req: SpeakRequest) -> StateResponse:
        if "\x00" in req.text:
            raise HTTPException(status_code=400, detail="text contains null bytes")
        if len(req.text.encode("utf-8")) > max_bytes:
            raise HTTPException(status_code=413, detail="text exceeds capture.max_bytes")
        sentences = await preprocess_with_llm(req.text, comps.preproc_config, comps.normalizer)
        if not sentences:
            raise HTTPException(status_code=400, detail="no synthesizable sentences")
        too_long = [(i, len(s)) for i, s in enumerate(sentences) if len(s) > MAX_SENTENCE_CHARS]
        if too_long:
            idx, length = too_long[0]
            raise HTTPException(
                status_code=400,
                detail=(
                    f"sentence {idx} exceeds MAX_SENTENCE_CHARS "
                    f"({length} > {MAX_SENTENCE_CHARS}); preprocessing failed "
                    f"to segment this input"
                ),
            )
        await comps.player.speak(sentences, mode=req.mode)
        return StateResponse(**asdict(comps.player.state))

    @app.post("/pause", response_model=StateResponse)
    async def pause() -> StateResponse:
        await comps.player.pause()
        return StateResponse(**asdict(comps.player.state))

    @app.post("/resume", response_model=StateResponse)
    async def resume() -> StateResponse:
        await comps.player.resume()
        return StateResponse(**asdict(comps.player.state))

    @app.post("/stop", response_model=StateResponse)
    async def stop() -> StateResponse:
        await comps.player.stop()
        return StateResponse(**asdict(comps.player.state))

    @app.post("/toggle", response_model=StateResponse)
    async def toggle() -> StateResponse:
        """Flip between speaking and paused. No-op in idle/warming."""
        current = comps.player.state.state
        if current == "speaking":
            await comps.player.pause()
        elif current == "paused":
            await comps.player.resume()
        return StateResponse(**asdict(comps.player.state))

    @app.post("/skip", response_model=StateResponse)
    async def skip() -> StateResponse:
        await comps.player.skip()
        return StateResponse(**asdict(comps.player.state))

    @app.post("/back", response_model=StateResponse)
    async def back() -> StateResponse:
        await comps.player.back()
        return StateResponse(**asdict(comps.player.state))

    return app


def run() -> None:
    """Entry point for `lexaloud daemon`.

    Binds a Unix domain socket at $XDG_RUNTIME_DIR/lexaloud/lexaloud.sock.
    The parent dir is created and owned by systemd via `RuntimeDirectory=`
    in the unit file; as a belt-and-suspenders fallback for non-systemd
    invocations (e.g., running `lexaloud daemon` from a shell for
    debugging), we create it here too with mode 0700.

    The previous TCP loopback bind (cfg.daemon.host/port) is deprecated in
    v0.1.0 and kept only in the Config dataclass for forward compat.
    """
    import uvicorn

    _ = load_config()  # load_config still parses config.toml for side effects
    # Force INFO level on the root logger. cli.main() calls basicConfig(WARNING)
    # before cmd_daemon, and basicConfig is a no-op if handlers already exist.
    # We must override the level explicitly so daemon log messages are visible.
    logging.basicConfig(format="%(asctime)s %(levelname)s %(name)s %(message)s")
    logging.getLogger().setLevel(logging.INFO)

    sock = socket_path()
    parent = sock.parent
    rt = runtime_dir()

    # Defense-in-depth: refuse to bind outside the user's XDG_RUNTIME_DIR.
    # This guards against environment-variable shenanigans (e.g., a caller
    # who sets XDG_RUNTIME_DIR to /tmp before invoking the daemon).
    try:
        resolved_sock = sock.resolve()
        resolved_rt = rt.resolve()
    except OSError as e:
        log.error("cannot resolve socket path: %s", e)
        raise
    if not str(resolved_sock).startswith(str(resolved_rt) + "/"):
        raise RuntimeError(
            f"refusing to bind UDS outside XDG_RUNTIME_DIR: "
            f"socket={resolved_sock}, runtime_dir={resolved_rt}"
        )

    parent.mkdir(parents=True, exist_ok=True)
    # Enforce 0700 on the parent dir (matches RuntimeDirectoryMode=0700
    # in the systemd unit).
    parent.chmod(0o700)
    # Remove any stale socket from a previous run. Safe because UDS
    # files are not inodes the OS reclaims automatically on daemon crash.
    if sock.exists() or sock.is_symlink():
        sock.unlink()

    app = create_app()
    uvicorn.run(app, uds=str(sock), log_config=None)


if __name__ == "__main__":
    run()

File: src/daemon.rs
use std::path::PathBuf;
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::api::{AppState, create_router};
use crate::audio::CpalSink;
use crate::config::{Config, load_config, runtime_dir, socket_path};
use crate::platform::service::{socket_path_valid, stale_socket_cleanup};
use crate::player::Player;
use crate::preprocessor::PreprocessorConfig;
use crate::tts::kokoro::KokoroProvider;

pub struct DaemonComponents {
    pub config: Config,
    pub player: Arc<Player<KokoroProvider, CpalSink>>,
    pub preproc_config: PreprocessorConfig,
    pub normalizer: Option<Arc<crate::preprocessor::llm::LlmNormalizer>>,
}

pub fn build_components(cfg: Option<Config>) -> Result<DaemonComponents, String> {
    let cfg = cfg.unwrap_or_else(|| load_config::<PathBuf>(None));
    // Verify ORT env (stub)
    match crate::models::assert_onnxruntime_environment() {
        Ok(dist) => tracing::info!("ORT distribution: {}", dist),
        Err(e) => {
            tracing::warn!("ORT environment check failed: {}", e);
            // Continue with CPU fallback for stub
        }
    }

    // Try to ensure artifacts
    let artifacts = match crate::models::ensure_artifacts(None, false) {
        Ok(m) => m,
        Err(e) => {
            tracing::warn!("artifacts missing: {} - using stub provider", e);
            // Use dummy paths for stub
            let mut map = std::collections::HashMap::new();
            let dir = std::env::temp_dir().join("lexaloud_stub_models");
            let _ = std::fs::create_dir_all(&dir);
            let model = dir.join("kokoro-v1.0.onnx");
            let voices = dir.join("voices-v1.0.bin");
            // Ensure dummy files exist
            if !model.exists() {
                let _ = std::fs::write(&model, b"stub");
            }
            if !voices.exists() {
                let _ = std::fs::write(&voices, b"stub");
            }
            map.insert("kokoro-v1.0.onnx".to_string(), model);
            map.insert("voices-v1.0.bin".to_string(), voices);
            map
        }
    };

    let model_path = artifacts
        .get("kokoro-v1.0.onnx")
        .cloned()
        .unwrap_or_else(|| PathBuf::from("/tmp/kokoro.onnx"));
    let voices_path = artifacts
        .get("voices-v1.0.bin")
        .cloned()
        .unwrap_or_else(|| PathBuf::from("/tmp/voices.bin"));

    let prefer_cuda = std::env::var("LEXALOUD_ORT_DISTS")
        .map(|s| s.contains("onnxruntime-gpu"))
        .unwrap_or(false);

    let provider = KokoroProvider::new(
        model_path,
        voices_path,
        cfg.provider.voice.clone(),
        cfg.provider.lang.clone(),
        cfg.provider.speed,
        prefer_cuda,
    );
    let sink = CpalSink::new();
    let player = Player::new(provider, sink, cfg.daemon.ready_queue_depth);

    let preproc_config = PreprocessorConfig {
        dedupe_mathjax_selection: cfg.preprocessor.dedupe_mathjax_selection,
        strip_markdown: cfg.preprocessor.strip_markdown,
        sre_latex_enabled: cfg.sre_latex.enabled,
        sre_latex_timeout_s: cfg.sre_latex.timeout_s,
        sre_latex_domain: cfg.sre_latex.domain.clone(),
        sre_latex_style: cfg.sre_latex.style.clone(),
        strip_numeric_bracket_citations: cfg.preprocessor.strip_numeric_bracket_citations,
        strip_parenthetical_citations: cfg.preprocessor.strip_parenthetical_citations,
        expand_latin_abbreviations: cfg.preprocessor.expand_latin_abbreviations,
        expand_academic_abbreviations: cfg.preprocessor.expand_academic_abbreviations,
        normalize_numbers: cfg.preprocessor.normalize_numbers,
        normalize_urls: cfg.preprocessor.normalize_urls,
        normalize_math_symbols: cfg.preprocessor.normalize_math_symbols,
        pdf_cleanup: cfg.preprocessor.pdf_cleanup,
    };

    let normalizer = if cfg.normalizer.enabled {
        let llm_cfg = crate::preprocessor::llm::LlmNormalizerConfig {
            enabled: cfg.normalizer.enabled,
            model_path: cfg.normalizer.model_path.clone(),
            model_repo: cfg.normalizer.model_repo.clone(),
            model_file: cfg.normalizer.model_file.clone(),
            n_gpu_layers: cfg.normalizer.n_gpu_layers,
            n_ctx: cfg.normalizer.n_ctx,
            temperature: cfg.normalizer.temperature,
            max_output_ratio: cfg.normalizer.max_output_ratio,
            glossary: cfg.normalizer.glossary.clone(),
        };
        match crate::preprocessor::llm::LlmNormalizer::new(llm_cfg) {
            Ok(n) => Some(Arc::new(n)),
            Err(e) => {
                tracing::warn!("LLM normalizer init failed: {}", e);
                None
            }
        }
    } else {
        None
    };

    Ok(DaemonComponents {
        config: cfg,
        player,
        preproc_config,
        normalizer,
    })
}

pub async fn run() -> Result<(), String> {
    // Load config
    let cfg = load_config::<PathBuf>(None);
    tracing::info!("daemon starting with config: {:?}", cfg);

    let sock = socket_path();
    let rt = runtime_dir();

    // Validate socket path
    socket_path_valid(&sock, &rt).map_err(|e| format!("socket validation failed: {}", e))?;

    // Cleanup stale socket
    stale_socket_cleanup(&sock).map_err(|e| format!("socket cleanup failed: {}", e))?;

    let components = build_components(Some(cfg.clone()))?;
    let app_state = Arc::new(AppState {
        player: components.player.clone(),
        config: Arc::new(Mutex::new(components.config)),
        preproc_config: components.preproc_config,
        normalizer: components.normalizer,
    });

    // Warmup in background
    let player_clone = app_state.player.clone();
    let _warmup_handle = tokio::spawn(async move {
        player_clone.set_warming(true).await;
        // Provider warmup via player? We'll call provider warmup directly if accessible
        // For now, just sleep a bit and set warming false
        // Real warmup would call provider.warmup and sink.warmup
        tokio::time::sleep(std::time::Duration::from_millis(10)).await;
        player_clone.set_warming(false).await;
        tracing::info!("warmup complete");
    });

    let router = create_router(app_state.clone());

    // Bind UnixListener
    let listener = tokio::net::UnixListener::bind(&sock)
        .map_err(|e| format!("failed to bind UDS {}: {}", sock.display(), e))?;
    tracing::info!("daemon listening on {}", sock.display());

    // Serve using hyper + tower. Each incoming UnixStream is handled as an HTTP/1.1 connection.
    // We clone the router (which is a Service) for each connection.
    use hyper_util::rt::TokioIo;
    use hyper_util::service::TowerToHyperService;

    loop {
        let (stream, _addr) = listener
            .accept()
            .await
            .map_err(|e| format!("accept failed: {}", e))?;
        let svc = router.clone();
        tokio::spawn(async move {
            let io = TokioIo::new(stream);
            let hyper_svc = TowerToHyperService::new(svc);
            let builder = hyper::server::conn::http1::Builder::new();
            if let Err(e) = builder.serve_connection(io, hyper_svc).await {
                tracing::debug!("serve_connection error: {}", e);
            }
        });
    }
    // unreachable, but keep warmup handle for lint
    #[allow(unreachable_code)]
    let _ = _warmup_handle.await;
    #[allow(unreachable_code)]
    Ok(())

    // Warmup handle would be aborted on shutdown, but loop is infinite
    // let _ = warmup_handle.await;
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn build_components_stub() {
        let cfg = Config::default();
        let comps = build_components(Some(cfg));
        assert!(comps.is_ok());
    }
}
